File: Subsets/Subsets.py
# 48ms 42.54%
class Solution:
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def dfs(nums, temp_list):
            for index in range(len(nums)):
                res_list.append(temp_list + [nums[index]])
                dfs(nums[index + 1:], temp_list + [nums[index]])
            
        # 这里因此res_list至少有一个[],所以提前先加好
        res_list = [[]]
        dfs(nums, [])
        return res_list

File: Subsets/Subsets.cpp
// 典型的深度优先搜索

// Runtime: 8 ms, faster than 100.00% of C++ online submissions for Subsets.
// Memory Usage: 9 MB, less than 91.46% of C++ online submissions for Subsets.

class Solution 
{
public:
    vector<vector<int>> subsets(vector<int>& nums) 
    {
        vector<vector<int>> res;
        vector<int> candidate;
        helper(res, nums, 0, candidate);
        return res;
    }
private:
    void helper(vector<vector<int>>& res, vector<int>& nums, int n, vector<int>& candidate)
    {
        res.push_back(candidate);
        for (int i = n; i < nums.size(); i++)
        {
            candidate.push_back(nums[i]);
            helper(res, nums, i + 1, candidate);
            candidate.pop_back();
        }
    }
};